### PR TITLE
Improved Docker image + publish on `ghcr.io`

### LINKED ISSUE #610: Add docker release images
It will be good to build and push on each release an image to dockerhub.

### PR DESCRIPTION
**Pull request by AFCMS**
Fix #610

- [x] Native cross-compilation support in Dockerfile. Only `linux/amd64` and `linux/arm64` supported and tested (Tier 1 platform support from Rust)
- [x] Cache support for cargo downloads and compilation results in Dockerfile
- [x] Open Container's [annotations](https://github.com/opencontainers/image-spec/blob/main/annotations.md) in Dockerfile 
- [x] GitHub Actions workflow to build the image for both platforms, publishing to ghcr.io on tags and master branch pushes.
- [x] Disable use of GitHub Actions cache for tags build, allow manually triggering the workflow with or without cache.
- [x] [Attestation artifacts](https://docs.github.com/en/actions/security-for-github-actions/using-artifact-attestations/using-artifact-attestations-to-establish-provenance-for-builds) for builds
- [x] Add basic informations about the use of the Docker image in README

I also changed the WORKDIR from `/src` to `/work`, because if got me confused with the use of the src folder for the project source in the first stage of the Dockerfile. It doesn't impact anything anyways since you can mount the file where you want and have the program options point to it.

**How to test?**
_After checking out the branch and making sure you have QEMU installed to build/test_
```shell
# Docker sadly doesn't have yet the way to have locally a tag with multiple platforms
# despite being able to pass multiple platforms to the build command, so we have to use two distinct tags.
docker build --platform=linux/amd64 --tag test-oxipng-amd:latest --load .
docker build --platform=linux/arm64 --tag test-oxipng-arm:latest --load .

docker run --rm test-oxipng-amd:latest --version
docker run --rm test-oxipng-arm:latest --version

# Run on some files
docker run --rm -it -v $(pwd):/work test-oxipng-amd:latest -a /work/tests/files/apng_file.png
docker run --rm -it -v $(pwd):/work test-oxipng-arm:latest -a /work/tests/files/apng_file.png

# Remove the images
docker image rm test-oxipng-amd:latest
docker image rm test-oxipng-arm:latest
```

For the workflow, see the GitHub Actions logs. If you want to test the ghcr.io publishing you can merge the branch into a fork and see the result.

### CODE REVIEW

**Review by AlexTMjugador — APPROVED** [review]
Looks good to me! I'd merge this after hearing @shssoichiro's thoughts about setting up the GitHub container registry for this repository. Great work on the documentation, Dockerfile and workflow. Thank you.

That said, as I mentioned in #610, I'm curious about what value these Docker images provide over the portable `musl` binaries. I understand that Docker images are often requested for CLI tools, and that’s fine, but _why exactly_ in this case? What specific problem does this solve for you?

Edit: I've just rebased your PR on top of the latest mainline commits to fix unrelated Clippy lint issues, please don't mind it :+1:

**Comment by AFCMS** [comment]
I don't have a really specific usecase, I did this PR mostly as a training for both Docker and Rust :sweat_smile: (I use the tool with `cargo install`)

Anyways, I think the benefit of a Docker container is that you don't need to go to the download page, extract the binary and then go back regularly to check if there is any new version.
With Docker you can just run the command in documentation with the latest image and the `--pull=always` flag, so it will download or update if needed and run the program automatically.

**Comment by AlexTMjugador** [comment]
Thanks for the comment @AFCMS, I do indeed realize now that the convenience of unifying software distribution through Docker containers can be valuable. Let's wait a bit for Josh's view on using GitHub Packages before merging :rocket:

**Comment by andrews05** [comment]
I wonder if it would be better for someone to maintain this externally from the main repo...

**Comment by AlexTMjugador** [comment]
> I wonder if it would be better for someone to maintain this externally from the main repo...

I think it's more discoverable and official-looking to use the same repository for Docker images and development, but anyone is free to create a repository for hosting their own OxiPNG Docker images of course.

**Comment by AlexTMjugador** [comment]
@andrews05 and I would like to move forward with a new Oxipng release. Since Josh didn't share any feedback on this PR after some time, I'm assuming there are no objections there, so I'll go ahead and merge it. Let’s see how it works out! :rocket: :crossed_fingers:

**Comment by AlexTMjugador** [comment]
Looks like this worked on the first time for pushes at least, great! :100:

[Oxipng container page](https://github.com/shssoichiro/oxipng/pkgs/container/oxipng)

**Comment by AFCMS** [comment]
I through about a even better usecase for the existence of a Docker image soon after this PR got merged, ironically on my own projects 😅

It makes it extremely easy to add a specific version of the oxipng binary in another container, with seamless integration with systems like Dependabot, instead of a multi step process with downloading a release archive or doing a custom build.

Anyways, thanks again for this amazing tool! It's one of my indispensable dev tools now.

**Pull request merged**
